uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequestCalendarFeedConfig.php

```diff
@@ -0,0 +1,115 @@
+<?php
+
+use CRM_HRLeaveAndAbsences_Exception_InvalidLeaveRequestCalendarFeedConfigException as InvalidLeaveRequestCalendarFeedConfigException;
+
+class CRM_HRLeaveAndAbsences_BAO_LeaveRequestCalendarFeedConfig extends CRM_HRLeaveAndAbsences_DAO_LeaveRequestCalendarFeedConfig {
+
+  /**
+   * Create a new LeaveRequestCalendarFeedConfig based on array-data
+   *
+   * @param array $params key-value pairs
+   *
+   * @return CRM_HRLeaveAndAbsences_DAO_LeaveRequestCalendarFeedConfig|NULL
+   */
+  public static function create($params) {
+    $entityName = 'LeaveRequestCalendarFeedConfig';
+    $hook = empty($params['id']) ? 'create' : 'edit';
+
+    CRM_Utils_Hook::pre($hook, $entityName, CRM_Utils_Array::value('id', $params), $params);
+    self::validateParams($params);
+    self::setDefaultParameterValues($params);
+
+    $instance = new self();
+    $instance->copyValues($params);
+    $instance->save();
+    CRM_Utils_Hook::post($hook, $entityName, $instance->id, $instance);
+
+    return $instance;
+  }
+
+  /**
+   * A method for validating the params passed to the
+   * LeaveRequestCalendarFeedConfig create method
+   *
+   * @param array $params
+   *   The params array received by the create method
+   *
+   * @throws InvalidLeaveRequestCalendarFeedConfigException
+   */
+  public static function validateParams($params) {
+    self::validateLeaveRequestCalendarFeedConfigTitle($params);
+    self::validateTimezone($params);
+  }
+
+
+  /**
+   * Validates that the Title of the Leave calendar feed configuration
+   * is unique.
+   *
+   * @param array $params
+   *
+   * @throws InvalidLeaveRequestCalendarFeedConfigException
+   */
+  private static function validateLeaveRequestCalendarFeedConfigTitle($params) {
+    $title = CRM_Utils_Array::value('title', $params);
+
+    if (!$title) {
+      return;
+    }
+
+    $leaveRequestCalendarFeedConfig = new self();
+    $leaveRequestCalendarFeedConfig->title = $title;
+
+    if (!empty($params['id'])) {
+      $id = (int) $params['id'];
+      $leaveRequestCalendarFeedConfig->whereAdd("id <> $id");
+    }
+
+    $leaveRequestCalendarFeedConfig->find(true);
+
+    if ($leaveRequestCalendarFeedConfig->id) {
+      throw new InvalidLeaveRequestCalendarFeedConfigException(
+        'A leave request calendar feed configuration with same title already exists!'
+      );
+    }
+  }
+
+  /**
+   * Validates that the timezone supplied for the feed configuration
+   * is a valid one.
+   *
+   * @param array $params
+   *
+   * @throws InvalidLeaveRequestCalendarFeedConfigException
+   */
+  private static function validateTimezone($params) {
+    $timeZone = CRM_Utils_Array::value('timezone', $params);
+
+    if (!$timeZone) {
+      return;
+    }
+    $timeZoneList = DateTimeZone::listIdentifiers(DateTimeZone::ALL);
```

Comment:
This extension already has a [validator for dates](https://github.com/compucorp/civihr/blob/1e304cfe35788547e75c09fc3678fa3fa4dac5a7/uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Validator/Date.php) and I think that validating timezones is not really the responsibility of a Calendar Feed. What do you think of adding a validator for timezones?

Reply:
I agree its a good Idea.